rust/rpc/src/lib.rs

```diff
@@ -184,58 +185,80 @@ impl<W: Write + Send> RpcLoop<W> {
+    /// Calls to the handler happen on the caller's thread.
     ///
-    /// Calls to the handler are guaranteed to preserve the order as they appear on
-    /// on the channel. At the moment, there is no way for there to be more than one
-    /// incoming request to be outstanding.
+    /// Calls to the handler are guaranteed to preserve the order as
+    /// they appear on on the channel. At the moment, there is no way
+    /// for there to be more than one incoming request to be outstanding.
     ///
     /// This method returns when the input channel is closed.
-    pub fn mainloop<R: BufRead, RF: Send + FnOnce() -> R>(&mut self,
-            rf: RF,
-            handler: &mut Handler) {
+    pub fn mainloop<'a, R, RF, H>(&mut self, rf: RF, handler: &mut H)
+    where R: BufRead,
+          RF: Send + FnOnce() -> R,
+          H: Handler,
+    {
+
         crossbeam::scope(|scope| {
             let peer = self.get_raw_peer();
             scope.spawn(move|| {
                 let mut reader = rf();
-                while let Some(json_result) = self.read_json(&mut reader) {
-                    match json_result {
-                        Ok(json) => {
-                            let is_method = json.as_object().map_or(false, |dict|
-                                dict.contains_key("method"));
-                            if is_method {
-                                self.peer.put_rx(json);
-                            } else {
-                                self.peer.handle_response(json);
+                loop {
+                    let json = match self.read_json(&mut reader) {
+                        Ok(json) => json,
+                        Err(err) => {
+                            // if parsing fails, we print the error and exit.
+                            print_err!("{:?}", err);
+                            break
+                        }
+                    };
+                    if json.is_response() {
+                        let id = json.get_id().unwrap();
+                        match json.into_response() {
+                            Ok(resp) => {
+                                let resp = resp.map_err(Error::from);
+                                self.peer.handle_response(id, resp);
+                            }
+                            Err(msg) => {
+                                print_err!("failed to parse response: {}", msg);
+                                break
                             }
                         }
-                        Err(err) => print_err!("Error decoding json: {:?}", err)
+                    } else {
+                        self.peer.put_rx(Some(json));
                     }
                 }
-                self.peer.put_rx(Value::Null);
-                // TODO: send disconnect error to all pending
             });
+
             let mut idle = VecDeque::<usize>::new();
             loop {
+                // only do idle work if no pending messages
```

Comment:
This comment is restating the code. I might write "pending messages take priority over idle work"